### PR TITLE
GraphQL setup

### LINKED ISSUE #17: Install GraphQL to Rails app
Helpful articles:
 - https://graphql-ruby.org
 - https://mattboldt.com/2019/01/07/rails-and-graphql/
 - https://www.howtographql.com/graphql-ruby/0-introduction/

### LINKED ISSUE #19: Add dogs table and corresponding GraphQL schema
Add GraphQL schema to satisfy user story #2 (basic dog info).

### LINKED ISSUE #20: Add users table and corresponding GraphQL schema
Add GraphQL schema to satisfy user story #4 (basic owner info on dog profile).

### PR DESCRIPTION
**Pull request by chakeresa**
**Changes proposed in this pull request:**
 - Add `User` model & factory, `users` table. A user has_many dogs. 
 - Add `Dog` model & factory, `dogs` table. A dog belongs_to a user.
 - Add seed file with dogs and users
 - Add `graphql` & `graphiql-rails` gems (resolves #17)
 - Add `DogType` & `UserType` for GraphQL (resolves #19 & resolves #20)
 - Add `dog` `dogs` `user` & `users` query types (and tests thereof)

**The following checks have been completed:**
 - [x] Tested my new feature(s) as well as any feasible edge cases (if possible)
 - [x] Checked `coverage/index.html` - did not add any new code that's not covered by testing (if possible)
 - [x] Merged in the latest master to my branch with `git pull origin master` & resolved merge conflicts
 - [x] Ran `rails db:migrate`
 - [x] Ran the test suite - all tests are passing (or maybe skipped)
 - [x] Checked affected endpoints in Postman / GraphiQL
 - [ ] Updated README for changes (new endpoints, new gems, etc)

**Notes:**
 - Let's discuss documentation requirements for a GraphQL API at our first checkin

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -28,8 +28,10 @@ gem 'bootsnap', '>= 1.1.0', require: false
 # Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
```
Comment (alexander-mathieu) [inline Gemfile]: 👍

File: app/controllers/graphql_controller.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    variables = ensure_hash(params[:variables])
+    query = params[:query]
+    operation_name = params[:operationName]
+    context = {
+      # Query context goes here, for example:
+      # current_user: current_user,
+    }
+    result = CrowdhoundBeSchema.execute(query, variables: variables, context: context, operation_name: operation_name)
+    render json: result
+  rescue => e
```
Comment (alexander-mathieu) [inline app/controllers/graphql_controller.rb]: We'll have to hit the production server and see what we get back, it seems like GraphQL is handling an error in development differently than in production.

File: app/graphql/types/dog_type.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+module Types
```
Comment (alexander-mathieu) [inline app/graphql/types/dog_type.rb]: 👍

File: app/graphql/types/user_type.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module Types
```
Comment (alexander-mathieu) [inline app/graphql/types/user_type.rb]: 👍

File: config/routes.rb [hunk 5]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
```
Comment (alexander-mathieu) [inline config/routes.rb]: It's great there are two separate routes for /graphql and /graphiql, better than toggling back and forth!

File: spec/rails_helper.rb [hunk 6]
```diff
@@ -1,3 +1,4 @@
+require('simplecov')
```
Comment (alexander-mathieu) [inline spec/rails_helper.rb]: So bizarre that this is line 1 is necessary here!

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
```
Comment (alexander-mathieu) [inline spec/models/user_spec.rb]: 👍

File: spec/models/dog_spec.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
```
Comment (alexander-mathieu) [inline spec/models/dog_spec.rb]: 👍

File: app/graphql/types/query_type.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
+module Types
```
Comment (alexander-mathieu) [inline app/graphql/types/query_type.rb]: I'll be working in this file this evening, getting the appropriate filters added for dog(s) / user(s).

File: spec/factories/users.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
```
Comment (alexander-mathieu) [inline spec/factories/users.rb]: Hipster sentences / paragraphs are the best kind.

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  has_many :dogs
+
+  validates_presence_of :first_name, :last_name, :email
+  validates :email, uniqueness: { case_sensitive: false }
```
Comment (alexander-mathieu) [inline app/models/user.rb]: Didn't know { case_sensitive } was a thing.

File: db/migrate/20190902172445_create_users.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
```
Comment (alexander-mathieu) [inline db/migrate/20190902172445_create_users.rb]: Glad we pulled out the locations / identities to their own tables!
